### PR TITLE
Sale Product configurator

### PR DESCRIPTION
**Pull request by FrancoisGe**

### CODE REVIEW

File: addons/sale_product_configurator/static/src/js/sale_product_field.js [hunk 1]
```diff
@@ -6,6 +6,33 @@ import { SaleOrderLineProductField } from '@sale/js/sale_product_field';
...
+                default_custom_product_template_attribute_value_id: selectedCustomPTAV.id,
+                default_custom_value: ptal.customValue,
+            });
+        };
+    }
+    await record.data.product_custom_attribute_value_ids.createAndReplace(contextRecords);
+
+    const noVariantPTAVIds = product.attribute_lines.filter(
+        ptal => ptal.create_variant === "no_variant" && ptal.attribute_values.length > 1
+    ).map(ptal => { return ptal.selected_attribute_value_id });
```
Comment (FrancoisGe) [inline addons/sale_product_configurator/static/src/js/sale_product_field.js]: remove return

File: addons/web/static/src/views/relational_model/static_list.js [hunk 2]
```diff
@@ -121,20 +121,52 @@ export class StaticList extends DataPoint {
...
+                await this._onUpdate();
+            }
+        });
+    }
+
+    async applyCommands(commands) {
+        await this._applyCommands(commands);
+        await this._onUpdate();
+    }
+
```
Comment (FrancoisGe) [inline addons/web/static/src/views/relational_model/static_list.js]: remove

File: addons/web/static/src/views/relational_model/static_list.js [hunk 3]
```diff
@@ -121,20 +121,52 @@ export class StaticList extends DataPoint {
...
                 manuallyAdded: true,
+                mode,
             });
             await this._addRecord(record, { position });
             await this._onUpdate({ withoutOnchange: !record._checkValidity({ silent: true }) });
             return record;
         });
     }
 
+    createAndReplace(contextRecords, noUpdate) {
```
Comment (FrancoisGe) [inline addons/web/static/src/views/relational_model/static_list.js]: Add comment: Maybe To remove we need it only for product configurator

File: addons/sale_product_configurator/static/src/js/sale_product_field.js [hunk 4]
```diff
@@ -98,18 +128,20 @@ patch(SaleOrderLineProductField.prototype, 'sale_product_configurator', {
...
+
                 this._onProductUpdate();
+                saleOrderRecord.data.order_line.leaveEditMode();
                 for (const optionalProduct of optionalProducts) {
                     const line = await saleOrderRecord.data.order_line.addNewRecord({
                         position: 'bottom',
+                        mode: "readonly",
                     });
-                    line.update(optionalProduct);
+                    applyProduct(line, optionalProduct);
```
Comment (aab-odoo) [inline addons/sale_product_configurator/static/src/js/sale_product_field.js]: we don't wait here?

File: addons/web/static/src/views/relational_model/static_list.js [hunk 5]
```diff
@@ -121,20 +121,49 @@ export class StaticList extends DataPoint {
...
 
+    /**
+     * TODO: We should probably delete this function.
+     * It is only used for the product configurator.
+     * It will take a list of contexts containing default
+     * values used to create the new records in the static list.
+     * It will then delete the old records from the static list
+     * and replace them with the new ones we have just created.
+     */
+    createAndReplace(contextRecords) {
```
Comment (aab-odoo) [inline addons/web/static/src/views/relational_model/static_list.js:151]: alphabetical order!!!!

**Review by aab-odoo — APPROVED** [review]

**Pull request merged**